File: Trie/Leetcode/Medium/AddAndSearchWord/AddAndSearchWord.cpp
// https://leetcode.com/problems/design-add-and-search-words-data-structure/
#include<bits/stdc++.h>
using namespace std;
class WordDictionary
{
private:
    class Node
    {
    public:
        Node *links[26];
        bool flag = false;

        bool contains(char ch)
        {
            return links[ch - 'a'] != nullptr;
        }

        void put(char ch)
        {
            links[ch - 'a'] = new Node();
        }

        Node *get(char ch)
        {
            return links[ch - 'a'];
        }

        void setEnd()
        {
            flag = true;
        }

        bool getEnd()
        {
            return flag;
        }
    };

    Node *root;

public:
    WordDictionary()
    {
        root = new Node();
    }

    void addWord(string word)
    {
        Node *node = root;
        for (int i = 0; i < word.length(); i++)
        {
            if (!node->contains(word[i]))
            {
                node->put(word[i]);
            }

            node = node->get(word[i]);
        }

        node->setEnd();
    }

    bool search(string &word, int i, Node *node)
    {
        if (i == word.length())
            return node->getEnd();

        if (word[i] == '.')
        {
            for (Node *curr : node->links)
            {
                if (curr != nullptr && search(word, i + 1, curr))
                {
                    return true;
                }
            }
            return false;
        }

        if (!node->contains(word[i]))
            return false;

        return search(word, i + 1, node->get(word[i]));
    }

    bool search(string word)
    {
        return search(word, 0, root);
    }
};

File: Trie/Leetcode/Medium/AddAndSearchWord/AddAndSearchWord.java
// https://leetcode.com/problems/design-add-and-search-words-data-structure/

public class AddAndSearchWord {
    class WordDictionary {

        class Node {

            Node links[];
            boolean flag;

            Node() {
                links = new Node[26];
                flag = false;
            }

            boolean contains(char ch) {
                return links[ch - 'a'] != null;
            }

            Node get(char ch) {
                return links[ch - 'a'];
            }

            void put(char ch) {
                links[ch - 'a'] = new Node();
            }

            void setEnd() {
                flag = true;
            }

            boolean getEnd() {
                return flag;
            }
        }

        Node root;

        public WordDictionary() {
            root = new Node();
        }

        public void addWord(String word) {
            Node node = root;
            for (int i = 0; i < word.length(); i++) {
                if (!node.contains(word.charAt(i))) {
                    node.put(word.charAt(i));
                }
                node = node.get(word.charAt(i));
            }

            node.setEnd();
        }

        boolean search(String word, int index, Node node) {
            if (index == word.length()) {
                return node.getEnd();
            }
            // match '.' with all characters possible
            if (word.charAt(index) == '.') {
                for (Node curr : node.links) {
                    if (curr != null && search(word, index + 1, curr)) {
                        return true;
                    }
                }
                return false;
            }

            if (!node.contains(word.charAt(index))) {
                return false;
            }

            return search(word, index + 1, node.get(word.charAt(index)));
        }

        public boolean search(String word) {
            return search(word, 0, root);
        }
    }
}
